### PR TITLE
Build Output Folder reorganization

### LINKED ISSUE #556: Organize output folder! aka Compile all Unit Tests and Utilities to subdirectories
## Summary
To clean up the clutter in the server output directory. This proposes we compile all unit tests and utilities to sub-directories.

### Proposed file structure
- :open_file_folder: *compile output folder*
  - :open_file_folder: data
  - :open_file_folder: default_templates
    - :open_file_folder: default_dbs
    - settings_template.cfg
  - :open_file_folder: logs
  - :open_file_folder: maps
  - :open_file_folder: scripts
  - :open_file_folder: utilities
    - :open_file_folder: mapviewer_data
    - :open_file_folder: unit-tests
      - AuthDBEventRegistry
      - AuthEventRegistry
      - CommonEventRegistry
      - CRUDPEventRegistry
      - event_gen
      - FriendshipEventRegistry
      - FriendshipServiceFunctions
      - GameDBEventRegistry
      - GameEventRegistry
      - ManifestBuilder
      - MapEventRegistry
    - AuthExerciser
    - binConverter
    - dbtool
    - mapviewer
    - piggtool
    - SEGS.exe
    - slav
    - texConverter
  - authserver
  - SEGSAdmin
  - segs.db
  - segs_game.db
  - settings.cfg
  - output.log
  - README.md

### PR DESCRIPTION
**Pull request by broxen**
This PR aims to reorganized the build output folder, as it was getting unruly. It creates some helper methods in Settings.cpp to handle finding the root folder regardless of what application is calling settings.

Also I did some minor code cleanup here and there.

Please TEST TEST TEST @Segs/testers 

### Issues closed
Closes #556 

### Additions/modifications proposed in this pull request:
- Organize the output of the out folder, and standardize our path names
- Give the db files an extension, so it's clear to the end user where to look if they need to edit manually (aka, reduce support question "where is my database file?")

### Proposed file structure
- :open_file_folder: *compile output folder*
  - :open_file_folder: data
  - :open_file_folder: default_setup
    - :open_file_folder: mysql
    - :open_file_folder: pgsql
    - :open_file_folder: sqlite
    - settings_template.cfg
  - :open_file_folder: logs
    - output.log
  - :open_file_folder: maps
  - :open_file_folder: scripts
  - :open_file_folder: utilities
    - :open_file_folder: mapviewer_data
    - :open_file_folder: unit-tests
      - AuthDBEventRegistry
      - AuthEventRegistry
      - CommonEventRegistry
      - CRUDPEventRegistry
      - event_gen
      - FriendshipEventRegistry
      - FriendshipServiceFunctions
      - GameDBEventRegistry
      - GameEventRegistry
      - ManifestBuilder
      - MapEventRegistry
    - AuthExerciser
    - binConverter
    - dbtool
    - mapviewer
    - piggtool
    - SEGS.exe
    - slav
    - texConverter
  - segs_server
  - SEGSAdmin
  - segs.db
  - segs_game.db
  - settings.cfg
  - README.md

### Notes
This was broken out of my original DBTool PR #778, which will remain focused on dbtool-v5

### CODE REVIEW

**Comment by broxen** [comment]
A positive side effect of this PR is that you can actually name your database files whatever you want. `segs_accounts.sql` or `segs_game_info.db` etc

**Comment by broxen** [comment]
Not sure why this took 19min to build =(
https://travis-ci.com/Segs/Segs/builds/107024079

**Comment by broxen** [comment]
Latest commit should fail due to:
`undefined reference to Settings::getSettingsPath()` etc in SEGSAdmin

Not sure why :man_shrugging: because we're including COMPONENTS and the correct header (Settings.h)

**Comment by broxen** [comment]
Herald did a lot of testing of this for me and we ran into some issues:

Win7 & Manjaro:
```
Checking for existing databases...
Setting arguments...
Failed to start DBTool Add User... **Please ensure you are NOT running as an administrator or sudo user**
Checking for existing databases...
WARNING: Not all databases were found. Please use the server setup to your left
```
With the following workflow
```
1. Setup CoH Data Files button not working properly in either environment
2. Extracting pigg Files progress bar fills up
3. Creating Default Directories just sits there at zero progress
a popup is saying
Stop Process? 
This process is taking longer than expected, do you want to wait?
```
Every time I click No it adds 10%
```
WARNING: We couldn't find the correct data and/or maps directories. Please use the server setup to your left
```

**Comment by broxen** [comment]
These results are odd because I have no issues locally on Fedora.

**Comment by broxen** [comment]
Latest commit resolves pathing issues in testing.

However there is a new issue that has come to light: `windepolyqt` doesn't correctly output the required libraries needed for SEGSAdmin and some of the utilities in `utilities` subfolder. If the user already has QT installed with components, then it appears as though the OS will find the required files and run properly.

Since we don't really want everyone to have to download QT, we need to correct the windeployqt issue and (preferably) stick all the required libraries in one location where they are easy to find `/libs` or something.

**Comment by broxen** [comment]
With enormous thanks to @miles200 @mobbyg and @HeraldOfOmega for testing. This should be ready for merge.

File: Projects/CoX/Common/Messages/Map/VisitMapCells.h [hunk 1]
```diff
@@ -11,52 +11,52 @@
...
-                for (uint16_t j = 0; j < 8; j++)
-                {
-                    if (m_visible_map_cells[i * 8 + j])
-                    {
-                        byte_sum += std::pow(2, j);
-                    }
-                }
-                cells_arr[i] =  byte_sum;
+                if (m_visible_map_cells[i * 8 + j])
+                    byte_sum += std::pow(2, j);
```
Comment (nemerle) [inline Projects/CoX/Common/Messages/Map/VisitMapCells.h]: Hmmmmmmmmm
this looks very much like a:
`byte_sum |= 1<<j;`

File: Projects/CoX/Servers/AuthServer/CMakeLists.txt [hunk 2]
```diff
@@ -60,6 +60,11 @@ IF(WIN32)
     include(Windeployqt)
     windeployqt(segs_server deploy)
 ENDIF()
+if(APPLE)
```
Comment (nemerle) [inline Projects/CoX/Servers/AuthServer/CMakeLists.txt]: not `elseif` ? :)

File: Components/Settings.h [hunk 3]
```diff
@@ -9,22 +9,28 @@
...
 
 private:
     Settings();
     Settings(Settings const&);
     Settings& operator=(Settings const&);
 
+    static QString      m_segs_dir;
     static QString      m_settings_path;
+    static QString      m_default_tpl_dir;
+    static QString      m_default_settings_path;
```
[the remark this answers is not in the document]
Comment (broxen) [reply Components/Settings.h]: doh, I missed this. Absolutely correct!

File: Components/Settings.cpp [hunk 4]
```diff
@@ -32,10 +39,11 @@ Settings::Settings()
 
 void Settings::setSettingsPath(const QString path)
 {
```
[the remark this answers is not in the document]
Comment (broxen) [reply Components/Settings.cpp]: Will correct

File: Projects/CoX/Common/Messages/Map/VisitMapCells.h (hunk 1 above)
Comment (broxen) [reply Projects/CoX/Common/Messages/Map/VisitMapCells.h]: Likely is. This was merely a formatting cleanup. Do you want me to patch that in a separate PR?

Comment (dracc) [reply Projects/CoX/Common/Messages/Map/VisitMapCells.h]: Okay, learning moment for me; How does `+= std::pow(2, j);` equals to `|= 1<<j;`?

I get the power-of-two part, but `|=` == `+=`? Or is the original logic wrong?

Comment (broxen) [reply Projects/CoX/Common/Messages/Map/VisitMapCells.h]: Filed PR #854

**Pull request merged**